extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPuller.java

```diff
@@ -345,6 +350,65 @@ public boolean apply(Throwable input)
     };
   }
 
+  private URI mostRecentInDir(final Path dir, final Pattern pattern) throws IOException
+  {
+    final PathFilter filter = new PathFilter()
+    {
+      @Override
+      public boolean accept(Path path)
+      {
+        return pattern == null || pattern.matcher(path.getName()).matches();
+      }
+    };
+    long modifiedTime = Long.MIN_VALUE;
+    URI mostRecentURI = null;
+    final FileSystem fs = dir.getFileSystem(config);
+    for (FileStatus status : fs.listStatus(dir, filter)) {
+      if (status.isFile()) {
+        final long thisModifiedTime = status.getModificationTime();
+        if (thisModifiedTime >= modifiedTime) {
+          modifiedTime = thisModifiedTime;
+          mostRecentURI = status.getPath().toUri();
+        }
+      }
+    }
+
+    return mostRecentURI;
+  }
+
+  /**
+   * Returns the latest modified file at the uri of interest.
+   * @param uri Either a directory or a file on HDFS. If it is a file, the parent directory will be searched.
+   * @param pattern A pattern matcher for file names in the directory of interest. Passing `null` results in matching any file in the directory.
+   * @return The URI of the file with the most recent modified timestamp.
+   */
+  @Override
+  public URI getLatestVersion(final URI uri, final Pattern pattern)
+  {
+    final Path path = new Path(uri);
+    try {
+      return RetryUtils.retry(
+          new Callable<URI>()
+          {
+            @Override
+            public URI call() throws Exception
+            {
+              final FileSystem fs = path.getFileSystem(config);
+              return mostRecentInDir(fs.isDirectory(path) ? path : path.getParent(), pattern);
+            }
+          },
+          shouldRetryPredicate(),
+          DEFAULT_RETRY_COUNT
+      );
+    }
+    catch (Exception e) {
+      if (e instanceof FileNotFoundException) {
+        return null;
```

Comment:
Should have had it as a catch statement anyways, I don't remember how this code snippet came about. Will change.